### PR TITLE
Validator

### PR DESCRIPTION
**Pull request by coreyadkins**

### CODE REVIEW

File: practice/validator/index.html [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        <h1>Identity Theft Sign-Up</h1>
+      </header>
+      <main>
+        <form action="submit" method="post">
+          <label>Name: <input id="name" type="text" name="name" placeholder="ex. Joe Joeson"/></label>
+          <label>Date of Birth: <input id="dob" type="text" name="DOB" placeholder="ex. 1934-04-31"/></label>
+          <label>Phone Number: <input id="phone" type="text" name="Phone Number" placeholder="ex. 123-123-1234"/></label>
+          <button name="submit">
+            Steal my identity!
+          </button>
```
Comment (echase6) [inline practice/validator/index.html:19]: Ha!

File: practice/validator/site.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+function getUserPhone() {
+  return $('#phone').val();
+}
+
+// 2. Transform Functions
+// No transform.
+
+// 3. Create
+var NAME_MATCH = /\w+\s\w+/;
+var DOB_MATCH = /\d\d\d\d-\d\d-\d\d/;
```
Comment (echase6) [inline practice/validator/site.js]: I believe \w will match a-z, A-Z and 0-9.  An alternative is [a-zA-Z] or [a-z] with an 'i' flag, making it case-insensitive.
\d\d\d\d can be replaced by \d{4}, the {4} indicating how many times the previous element is repeated.

File: practice/validator/site.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+var PHONE_MATCH = /\d\d\d-\d\d\d-\d\d\d\d/;
+/**
```
Comment (echase6) [inline practice/validator/site.js:30]: Similar to above.  To get even fancier, since \d\d\d- is repeated twice, this can be grouped by parentheses and repeated.  Something like (\d{3}-){2}

File: practice/validator/site.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+ */
+function validate(match, box, input) {
+  var doesMatch = match.test(input);
+  var alreadyInvalid = box.hasClass('invalid');
+  /**
+   * Ensures that an empty form will not be marked as invalid.
+   */
+  if (input === '') {
+    doesMatch = true;
+  }
```
Comment (echase6) [inline practice/validator/site.js]: This can be accomplished in the RegEx.  The entire matching pattern can be included in a (...)?, which matches it 0 or 1 times, i.e., nothing will provide a valid match.

File: practice/validator/site.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  if (doesMatch === false && alreadyInvalid === false) {
+    toggleInvalid(box);
+  }
+  if (doesMatch === true && alreadyInvalid === true) {
+    toggleInvalid(box);
+  }
```
Comment (echase6) [inline practice/validator/site.js:72]: I suppose this is a style thing, but the way I read this is that you are doing a lot of work to keep track of which way the invalid class gets applied.  How about using the attr() to set the invalid class and removeClass() to remove the invalid class (when valid), instead of toggling it?

File: practice/validator/site.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+// 5. Main
+/**
+ * Pipes variables to test the 'name' input box for validity.
+ */
+function validateName() {
+  var inputName = getUserName();
+  var box = getNameBox();
+  validate(NAME_MATCH, box, inputName);
+}
```
Comment (echase6) [inline practice/validator/site.js:83]: I like the use of a generic validator.  Nice!

File: practice/validator/site.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+// 6. Register Functions
+/**
+ * Registers when the user is typing and calls the appropriate function to
+ * check validity of input.
+ */
+function registerInitialEventHandlers() {
+  $('#name').on('input', validateName);
+  $('#dob').on('input', validateDOB);
+  $('#phone').on('input', validatePhone);
+  $('form').on('submit', function(event) {
```
Comment (echase6) [inline practice/validator/site.js:110]: I used 'keyup' events, but I see that 'input' is a better choice.  Nice find!

File: practice/validator/site.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+ */
+function getDOBBox() {
+  return $('#dob');
+}
+/**
+ * Selects the 'phone' input box.
+ */
+function getPhoneBox() {
+  return $('#phone');
+}
```
Comment (echase6) [inline practice/validator/site.js:47]: Since your id's are very descriptive (#name, #dob, #phone), I would not have any trouble understanding your code if you eliminated these functions and just used the $('#...') where the call was made.

File: practice/validator/site.js (hunk 3 above)
Comment (coreyadkins) [inline practice/validator/site.js:30]: Nice. Somebody was studying up on their RegEx

File: practice/validator/site.js (hunk 8 above)
Comment (coreyadkins) [inline practice/validator/site.js:47]: Yeah,  I get that they are totally unnecessary, I created these mostly to observe David's instruction to separate this kind of work out of main, where we just want it to be piping things through.

File: practice/validator/site.js (hunk 5 above)
Comment (coreyadkins) [inline practice/validator/site.js:72]: Good idea. Working with toggle was a bit of a pain in the ass, I think I just knew it was a function that did what I wanted but didn't consider the alternatives!

File: practice/validator/site.js (hunk 6 above)
Comment (coreyadkins) [inline practice/validator/site.js:83]: giphy

File: practice/validator/site.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+// 4. Modify and Sync
+/**
+ * Turns on the invalid class to mark that the input text is invalid.
+ */
+function addInvalid(box) {
+  box.attr('class', 'invalid');
```
Comment (davidselassie) [inline practice/validator/site.js]: Since it's possible that there might be other classes applied to this element. `.addClass('invalid')` handles all of the edge cases.

File: practice/validator/site.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  var doesMatch = match.test(input);
+  /**
+   * Ensures that an empty form will not be marked as invalid.
+   */
+  if (doesMatch === false) {
+    addInvalid(box);
+  }
+  if (doesMatch === true) {
+    removeInvalid(box);
+  }
```
Comment (davidselassie) [inline practice/validator/site.js:72]: To check if something is true, just assert itself in the if statement. Also if you want to gate on a boolean being true or false, it's concise and readable to just use if-else.

```
if (doesMatch) {
  ...
}
```
